src/io/utils.rs

```diff
@@ -104,15 +104,49 @@ pub trait Rotation {
 }
 
 impl Rotation for DMatrix<usize> {
-    fn rotate_90(&self) -> DMatrix<usize> {
-        assert_eq!(self.nrows(), self.ncols());
-        let side = self.nrows();
-        let mut target_matrix = DMatrix::<usize>::zeros(side, side);
+    fn rotate_90(&self) -> Self {
+        let iter = (0..self.nrows()).rev().flat_map(|i| {
+            (0..self.ncols()).map(move |j| self[(i, j)])
+        });
+        DMatrix::<usize>::from_iterator(self.ncols(), self.nrows(), iter)
+    }
+}
+
+pub trait Reflection {
+    fn reflect_vertical(&self) -> Self;
+
+    fn reflect_horizontal(&self) -> Self;
+}
 
-        (0..side).for_each(|i| {
-            (0..side).for_each(|j| target_matrix[(j, (side - 1) - i)] = self[(i, j)]);
+impl Reflection for DMatrix<usize> {
+    fn reflect_vertical(&self) -> Self {
+        let iter = (0..self.ncols()).flat_map(|j| {
+            (0..self.nrows()).rev().map(move |i| self[(i, j)])
         });
-        target_matrix
+        DMatrix::<usize>::from_iterator(self.nrows(), self.ncols(), iter)
+    }
+
+    fn reflect_horizontal(&self) -> Self {
+        let iter = (0..self.ncols()).rev().flat_map(|j| {
+            (0..self.nrows()).map(move |i| self[(i, j)])
+        });
+        DMatrix::<usize>::from_iterator(self.nrows(), self.ncols(), iter)
+    }
+}
+
+pub trait DiagonalReflection {
+    fn reflect_top_left(&self) -> Self;
+
+    fn reflect_bottom_left(&self) -> Self;
+}
+
+impl DiagonalReflection for DMatrix<usize> {
+    fn reflect_top_left(&self) -> Self {
+        self.rotate_90().reflect_vertical()
+    }
+
+    fn reflect_bottom_left(&self) -> Self {
+        self.rotate_90().reflect_horizontal()
```

Comment:
Does this need to be trait bound by rotation?

Reply:
Not in this case, mainly because we're not implementing with generics. The trait `DiagonalReflection` does not need to be bound, because the implementation of this trait does not need to depend on the `Rotation` trait. The impl of `DiagonalReflection` for `DMatrix<usize>` does not need to be bound because we have already provided an implementation of `Rotation` for the concrete type `DMatrix<usize>`, we can just use those functions directly.